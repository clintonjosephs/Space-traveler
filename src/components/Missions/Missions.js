import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import fetchMissions, {
  toggleMissionSuccesful,
} from '../../redux/missions/missions';
import styles from './Missions.module.css';

let fetched = false;

const Missions = () => {
  const missions = useSelector((state) => state.missionReducer.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!fetched) {
      dispatch(fetchMissions());
      fetched = true;
    }
  }, []);

  const handleClick = (id) => {
    dispatch(toggleMissionSuccesful(id));
  };

  return (
    <table>
      <tbody>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th> </th>
        </tr>
        {missions.map(
          ({
            mission_name: missionName,
            mission_id: missionId,
            description,
            reserved,
          }) => (
            <tr key={missionId}>
              <td className={styles['mission-name']}>{missionName}</td>
              <td className={styles['mission-description']}>{description}</td>
              <td className={styles['mission-status']}>
                {' '}
                <p
                  className={
                    reserved ? styles.reserved : styles['not-reserved']
                  }
                >
                  {reserved ? 'Active Member' : 'NOT A MEMBER'}
                </p>
              </td>
              <td>
                <button
                  id={missionId}
                  onClick={(e) => handleClick(e.target.id)}
                  className={
                    reserved
                      ? styles['mission-joined']
                      : styles['mission-not-joined']
                  }
                  type="button"
                >
                  {reserved ? 'Leave Mission' : 'Join Mission'}
                </button>
              </td>
            </tr>
          ),
        )}
      </tbody>
    </table>
  );
};

export default Missions;
