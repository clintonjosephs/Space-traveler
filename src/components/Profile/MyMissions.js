import PropTypes from 'prop-types';
import { BsFillTrashFill } from 'react-icons/bs';
import { useDispatch } from 'react-redux';
import { toggleMissionSuccesful } from '../../redux/missions/missions';
import styles from './Profile.module.css';

const MyMissions = ({ missions }) => {
  const dispatch = useDispatch();

  const handleClick = (id) => {
    dispatch(toggleMissionSuccesful(id));
  };

  return (
    <ul className={styles['my-journeys']}>
      <li>
        {' '}
        <h1>My Missions</h1>
      </li>
      {missions.length === 0 ? (
        <li>No Missions joined yet</li>
      ) : (
        missions.map(({ mission_id: missionId, mission_name: missionName }) => (
          <li key={missionId}>
            <div className={styles.setItems}>
              <span>{missionName}</span>
              <BsFillTrashFill onClick={() => handleClick(missionId)} />
            </div>
          </li>
        ))
      )}
    </ul>
  );
};

MyMissions.propTypes = {
  missions: PropTypes.arrayOf(
    PropTypes.shape({
      missionName: PropTypes.string,
      missionId: PropTypes.string,
    }),
  ).isRequired,
};

export default MyMissions;
