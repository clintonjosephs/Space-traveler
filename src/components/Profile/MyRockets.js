import PropTypes from 'prop-types';
import { BsFillTrashFill } from 'react-icons/bs';
import { useDispatch } from 'react-redux';
import styles from './Profile.module.css';
import { reserveRocket } from '../../redux/rockets/rockets';

const MyRockets = ({ data }) => {
  const dispatch = useDispatch();
  const removeRocketHandler = (id) => {
    dispatch(reserveRocket(id));
  };
  return (
    <ul className={styles['my-journeys']} data-testid="profileRockets">
      <li>
        {' '}
        <h1>My Rockets</h1>
      </li>
      {data.length === 0 ? (
        <li>No rocket reservations yet</li>
      ) : (
        data.map(({ id, rocketName }) => (
          <li key={id}>
            <div className={styles.setItems}>
              <span>{rocketName}</span>
              <BsFillTrashFill onClick={() => removeRocketHandler(id)} role="img" />
            </div>
          </li>
        ))
      )}
    </ul>
  );
};

MyRockets.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      rocketName: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      reserved: PropTypes.bool.isRequired,
    }),
  ).isRequired,
};

export default MyRockets;
