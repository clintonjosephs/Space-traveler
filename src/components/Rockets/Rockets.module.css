.listContainer {
  margin: 0.6rem 6.2rem;
}

.img {
  margin-right: 1rem !important;
  border-top-left-radius: 5px !important;
  border-bottom-left-radius: 5px !important;
}

.listItems {
  display: flex;
  flex-direction: row;
  margin-bottom: 2rem;
  box-shadow: 0 0 10px var(--neutral-color-2);
  border-radius: 5px;
  transition: all 0.2s ease-in-out;
}

.listItems:hover {
  transform: scale(1.05);
}

.listWrapper {
  padding: 0.8rem;
  line-height: 1.5rem;
}

.listWrapper div {
  font-size: 1.2rem;
  padding: 1rem 0 1rem;
}

.listWrapper button {
  width: 12rem;
  height: 3rem;
  box-shadow: none;
  font-size: 1rem;
  border-radius: 5px;
  cursor: pointer;
  margin: 0.2rem;
}

.reserve {
  background: var(--accent-color);
  color: var(--white);
  border: none;
}

.cancel {
  color: var(--neutral-color-2);
  background: var(--white);
  border: solid 3px var(--neutral-color);
}

.badge {
  background: var(--accent-color-2);
  color: var(--white);
  border-radius: 5px;
  padding: 5px;
  font-size: 0.8rem;
  margin-right: 0.3rem;
}

.describe {
  line-height: 2rem;
  text-align: justify;
}

@media (max-width: 786px) {
  .listContainer {
    margin: 0;
    padding: 0;
    width: 100%;
  }

  .img {
    height: 100%; 
    overflow: auto;
    width: 100% !important;
    margin-right: 1rem !important;
  }

  .listItems {
    display: flex;
    flex-direction: column;
  }
}
