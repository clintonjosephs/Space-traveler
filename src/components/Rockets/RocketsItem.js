import React from 'react';
import PropTypes from 'prop-types';
import styles from './Rockets.module.css';

const RocketsItem = (props) => {
  const {
    rocketId, rocketName, description, flickerImages,
  } = props;
  const clickHandler = () => {
    console.log(rocketId);
  };
  return (
    <li className={styles.listItems}>
      <img src={flickerImages} alt={rocketName} />
      <div className={styles.listWrapper}>
        <h1>{rocketName}</h1>
        <div>
          <span className={styles.badge}>Reserved</span>
          {description}
        </div>
        <button type="button" className={styles.reserve} onClick={clickHandler}>
          Reserve Rocket
        </button>
        <button type="button" className={styles.cancel} onClick={clickHandler}>
          Cancel Reservation
        </button>
      </div>
    </li>
  );
};

RocketsItem.propTypes = {
  rocketId: PropTypes.string.isRequired,
  rocketName: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  flickerImages: PropTypes.string.isRequired,
};

export default RocketsItem;
