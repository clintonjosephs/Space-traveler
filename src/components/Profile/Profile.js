import { useSelector } from 'react-redux';
import MyMissions from './MyMissions';
import MyRockets from './MyRockets';
import styles from './Profile.module.css';

const Profile = () => {
  const rocketState = useSelector(
    (state) => state.rocketReducer.rockets,
  ).filter((rocket) => rocket.reserved === true);

  return (
    <div className={styles.container}>
      <MyMissions />
      <MyRockets data={rocketState} />
    </div>
  );
};

export default Profile;
