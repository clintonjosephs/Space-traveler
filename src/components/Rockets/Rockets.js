import React from 'react';
import { useSelector } from 'react-redux';
import RocketsItem from './RocketsItem';
import styles from './Rockets.module.css';

const Rockets = () => {
  const data = useSelector((state) => state.rocketReducer.rockets);

  if (data.length === 0) {
    return (
      <>
        <ul className={styles.listContainer}>
          <li className={styles.listItems} style={{ padding: '2rem' }}><h1>No rockets available</h1></li>
        </ul>
      </>
    );
  }
  return (
    <>
      <ul className={styles.listContainer}>
        {data.map(({
          id, rocketName, description, flickrImages,
        }) => (
          <RocketsItem
            key={id}
            rocketName={rocketName}
            description={description}
            flickerImages={flickrImages}
            id={id}
          />
        ))}
      </ul>
    </>
  );
};

export default Rockets;
