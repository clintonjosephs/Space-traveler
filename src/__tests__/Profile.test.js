import React from 'react';
import { Provider } from 'react-redux';
import {
  render, screen, fireEvent, cleanup,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import store from '../redux/ConfigureStore';
import Profile from '../components/Profile/Profile';
import Rockets from '../components/Rockets/Rockets';
import Missions from '../components/Missions/Missions';

const RocketsProvider = () => (
  <Provider store={store}>
    <Rockets />
  </Provider>
);

const ProfileProvider = () => (
  <Provider store={store}>
    <Profile />
  </Provider>
);

const MissionsProvider = () => (
  <Provider store={store}>
    <Missions />
  </Provider>
);

afterEach(() => {
  cleanup();
});

describe('Check MyRockets', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<ProfileProvider />).toJSON();
    expect(component).toMatchSnapshot();
  });
});

describe('Rockets Component', () => {
  it('renders rockets and perform click of reservation button', async () => {
    render(<RocketsProvider />);
    const allRockets = await screen.findAllByText('Reserve Rocket');
    expect(allRockets).toHaveLength(4);

    const reserveButton = await screen.findAllByText(/Reserve Rocket/i);
    fireEvent.click(reserveButton[0]);
    fireEvent.click(reserveButton[1]);
    fireEvent.click(reserveButton[2]);
  });
});

describe('Run tests on profile page functions', () => {
  it('persists the reservation made on the rockets component', () => {
    render(<ProfileProvider />);
    const profileRockets = screen.getByTestId('profileRockets');
    expect(profileRockets.childElementCount).toBe(4);
  });

  it('removes the reserved rockets when clicked', async () => {
    render(<ProfileProvider />);
    const deleteButtons = await screen.findAllByRole('img');
    expect(deleteButtons.length).toBe(3);

    fireEvent.click(deleteButtons[0]);
    fireEvent.click(deleteButtons[1]);

    const profileRockets = screen.getByTestId('profileRockets');
    expect(profileRockets.childElementCount).toBe(2);
  });
});

describe('Profile component', () => {
  it('renders profile', async () => {
    render(<ProfileProvider />);
    const headingElement1 = await screen.findByText('My Missions');
    const headingElement2 = await screen.findByText('My Rockets');
    expect(headingElement1).toMatchSnapshot();
    expect(headingElement2).toMatchSnapshot();
  });

  it('renders no missions joined', async () => {
    render(<ProfileProvider />);
    const listedItem = await screen.findByText('No Missions joined yet');
    expect(listedItem).toMatchSnapshot();
  });

  it('renders missions and perform click of reservation button', async () => {
    render(<MissionsProvider />);
    const allMissions = await screen.findAllByText('Join Mission');
    expect(allMissions).toHaveLength(10);
    fireEvent.click(allMissions[0]);
    fireEvent.click(allMissions[1]);
    fireEvent.click(allMissions[2]);
  });

  describe('Run tests on profile page functions', () => {
    it('persists the reservation made on the missions component', async () => {
      render(<ProfileProvider />);
      const joinedMission1 = await screen.findByText('Thaicom');
      const joinedMission2 = await screen.findByText('Telstar');
      const joinedMission3 = await screen.findByText('Iridium NEXT');
      expect(joinedMission1).toMatchSnapshot();
      expect(joinedMission2).toMatchSnapshot();
      expect(joinedMission3).toMatchSnapshot();
    });

    it('removes the reserved missions when clicked', async () => {
      render(<ProfileProvider />);
      const deleteButtons = await screen.findAllByRole('figure');
      expect(deleteButtons.length).toBe(3);
      fireEvent.click(deleteButtons[0]);
      fireEvent.click(deleteButtons[1]);
      fireEvent.click(deleteButtons[2]);
      const listedItem = await screen.findByText('No Missions joined yet');
      expect(listedItem).toMatchSnapshot();
    });
  });
});
