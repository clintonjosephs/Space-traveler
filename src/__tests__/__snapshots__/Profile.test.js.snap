// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check MyRockets matches snapshot 1`] = `
<div
  className="container"
>
  <ul
    className="my-journeys"
  >
    <li>
       
      <h1>
        My Missions
      </h1>
    </li>
    <li>
      No Missions joined yet
    </li>
  </ul>
  <ul
    className="my-journeys"
    data-testid="profileRockets"
  >
    <li>
       
      <h1>
        My Rockets
      </h1>
    </li>
    <li>
      No rocket reservations yet
    </li>
  </ul>
</div>
`;
